test/test_timer.cpp

```diff
@@ -0,0 +1,75 @@
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE TimerTest
+#include <boost/test/unit_test.hpp>
+#include <clock.hpp>
+#include <thread>
+
+using namespace drumpi;
+using namespace clock;
+
+BOOST_AUTO_TEST_CASE(constructors) {
+    // Test for NULL object pointers
+    Timer t1;
+    Timer t2;
+
+    BOOST_CHECK(&t1);
+    BOOST_CHECK(&t2);
+}
+
+BOOST_AUTO_TEST_CASE(triggerTimeSetting) {
+    // Test that Timer delay is set properly
+    Timer t;
+    int d = 100; // delay = 100ms
+
+    t.setTime(d);
+
+    BOOST_CHECK_EQUAL(d, t.getTime());
+}
+
+class TestTimer1 : public Timer {
+    public:
+    int i = 0;
+    void trigger() override {
+        i++;
+    }
+};
+
+BOOST_AUTO_TEST_CASE(incrementOnce) {
+    // Test that the Timer triggers once and only once
+    TestTimer1 t;
+    int d = 10; // delay = 10ms
+    t.setTime(d);
+
+    BOOST_CHECK_EQUAL(0, t.i);
+
+    t.start();
+
+    BOOST_CHECK_EQUAL(0, t.i);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(d + (d/10)));
+
+    BOOST_CHECK_EQUAL(1, t.i);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(d * 2));
+
+    BOOST_CHECK_EQUAL(1, t.i);
+}
+
+BOOST_AUTO_TEST_CASE(stopTimer) {
+    // Test stopping the timer before the timer triggers
+    TestTimer1 t;
+    int d = 10; // delay = 10ms
+    t.setTime(d);
+
+    t.start();
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(d/2));
+
+    t.stop();
+
+    BOOST_CHECK_EQUAL(0, t.i);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(d));
+
+    BOOST_CHECK_EQUAL(0, t.i);
+}
```

Comment:
GitHub is recommending adding a new line at the end of files. Don't know if this will actually cause an issue or if just considered good practice?

Reply:
I think it's just good practice. If GitHub can do it automatically go for it, bit of a waste of time doing that manually I'd say though.

Reply:
Can't see a way to do it automatically. If it causes issues we can change it, for now no point wasting time changing it if it works.